Issue 649 node validation

melowe opened the pull request:
Validate nodes sending party info by providing validation callback. When nodes attempt to send party info check they are trustworthy but sending them some data to be unencrypted.

melowe committed:
Commit current code to remote for review

melowe committed:
Node keys are validated by sending them an encrypted string that 
the sending node unencrypt.

melowe committed:
Readd workaround for strange coverage issue with lambdas

melowe committed:
Merge branch 'master' into issue_649_node_validation

prd-fox committed:
Don't load resource file, instead use a stream for the bytes
Add extra debug info if needed

melowe committed:
Allow for multiple keys for a given url. 
Treat calling node not providing any valid keys as a security issue. 
Filter all recipients by valid key.

melowe committed:
Remove unessesary check on key - url pairs on storage all certs or 
prefiltered before this stage

melowe committed:
Lookup first key not default.

melowe committed:
Only validate certs for the node that is sending the data.

melowe committed:
Revert "Only validate certs for the node that is sending the data. "

This reverts commit b7e641b696e2c530610f1b81673cd4489ae82df5.

melowe committed:
Revert "Remove unessesary check on key - url pairs on storage all certs or  prefiltered before this stage "

This reverts commit bf2d77e1cc3dfcdd041e717a25747c704e5b0ab8.

prd-fox committed:
Don't rely on library to read the DDL

prd-fox committed:
Remove unused import

melowe committed:
Merge branch 'master' into issue_649_node_validation

melowe committed:
Merge origin/issue_649_node_validation into issue_649_node_validation

melowe committed:
Merge branch 'master' into issue_649_node_validation

melowe committed:
Merge origin/issue_649_node_validation into issue_649_node_validation

melowe committed:
Allow recipient url updates as we validate keys against the host before
 allowing them to connect.

melowe committed:
Add warning level message . Only reveal that a key has failed validation 
for the url and that the recipient will be ignored.

melowe committed:
Merge branch 'master' into issue_649_node_validation

melowe committed:
Merge branch 'master' into issue_649_node_validation

melowe committed:
Merge origin/fix-resource-reading into issue_649_node_validation

melowe committed:
Merge origin/issue_649_node_validation into issue_649_node_validation

prd-fox commented on jaxrs-service/src/main/java/com/quorum/tessera/p2p/PartyInfoResource.java:
```diff
@@ -1,55 +1,143 @@
+
+        final Set<Recipient> recipients = partyInfo.getRecipients()
+            .stream()
+            .filter(isValidRecipientKey)
```
We should first filter on `isSendingUrl` to only try to check keys the sending node claimed to own, otherwise we will get a lot of expected misses.

melowe replied:
That’s the first check. The second is creating a filtered list of all recipients. If we don’t need to validate the recipients in party info then yes we can collect the initial filters

melowe committed:
Update jaxrs-service/src/main/java/com/quorum/tessera/p2p/PartyInfoResource.java

Co-Authored-By: Peter Fox <31221824+prd-fox@users.noreply.github.com>

melowe committed:
Merge branch 'master' into issue_649_node_validation

melowe committed:
Merge origin/issue_649_node_validation into issue_649_node_validation

melowe committed:
Assume that any error during key validation renders the key invalid.
Add prefix to test suite annotations and executioncontext to allow
for custom namespaces when generating files and loggers. Add inital
tests for p2p changes to ensure interactions with 4 test nodes are
expected.
